pyro/optim/dct_adam.py

```diff
@@ -86,31 +101,36 @@ def step(self, closure=None):
 
                 # State initialization
                 if len(state) == 0:
-                    state['step'] = 0
+                    state['step'] = torch.zeros_like(p)
                     # Exponential moving average of gradient values
-                    state['exp_avg'] = torch.zeros_like(grad)
+                    state['exp_avg'] = torch.zeros_like(p)
                     # Exponential moving average of squared gradient values
-                    state['exp_avg_sq'] = torch.zeros_like(grad)
+                    state['exp_avg_sq'] = torch.zeros_like(p)
 
                 exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
                 beta1, beta2 = group['betas']
 
-                state['step'] += 1
+                state['step'][mask] += 1
 
                 # Decay the first and second moment running average coefficient
-                exp_avg.mul_(beta1).add_(1 - beta1, grad)
-                exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
+                exp_avg.masked_scatter_(mask, exp_avg[mask].mul_(beta1).add_(1 - beta1, grad))
+                exp_avg_sq.masked_scatter_(
+                    mask, exp_avg_sq[mask].mul_(beta2).addcmul_(1 - beta2, grad, grad))
 
-                denom = exp_avg_sq.sqrt().add_(group['eps'])
+                denom = exp_avg_sq[mask].sqrt().add_(group['eps'])
 
-                bias_correction1 = 1 - beta1 ** state['step']
-                bias_correction2 = 1 - beta2 ** state['step']
-                step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
+                bias_correction1 = 1 - beta1 ** state['step'][mask]
+                bias_correction2 = 1 - beta2 ** state['step'][mask]
+                step_size = group['lr'] * torch.sqrt(bias_correction2) / bias_correction1
 
                 if time_dim is None:
-                    p.data.addcdiv_(-step_size, exp_avg, denom)
+                    step = exp_avg[mask] / denom
+                    p.data[mask] += (step.mul_(-step_size))
                 else:
-                    step = _transform_inverse(exp_avg / denom, time_dim, duration)
-                    p.data.add_(step.mul_(-step_size))
+                    step = _transform_inverse(exp_avg[mask] / denom, time_dim, duration)
+                    p.data[mask] += (step.mul_(-step_size))
+
+                if hasattr(p, "_pyro_subsample"):
```

Comment:
why is this necessary?

Reply:
This is not necessarily required. I just want to make sure those extra attributes are cleared if, for some reason, users want to run SVI with subsampling at the early training phase and run SVI without subsampling at later steps. This might be not a good place to put that logic. Do you have other suggestions?